### ISSUE TITLE
[Beta] Add a sharded grad scaler in PyTorch distributed

### ISSUE THREAD

**Issue by zhaojuanmao** [pytorch/pytorch#64632]
FSDP needs a sharded grad scaler to scale gradients when there are overflows in mixed precision training. 

Fairscale version of sharded grad scaler does not work with CPU offload, we may need to fix that.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

**Comment by mrshenli** [comment]
`ZeroRedundancyOptimizer` probably does not need it. Because it is supposed to be used in conjunction with DDP, which will expose the access to full gradients.

**Comment by zhaojuanmao** [comment]
> `ZeroRedundancyOptimizer` probably does not need it. Because it is supposed to be used in conjunction with DDP, which will expose the access to full gradients.

I see, sounds good, let me update the description

**Comment by daniellepintz** [comment]
Hi @zhaojuanmao, what is the status of this issue?

**Comment by rohan-varma** [comment]
Thanks for pinging @daniellepintz  - We've shipped a sharded grad scaler: https://github.com/pytorch/pytorch/blob/master/torch/distributed/fsdp/sharded_grad_scaler.py
